src/wkt/src/message.rs

```diff
@@ -20,27 +20,52 @@ use crate::AnyError as Error;
 /// A trait that must be implemented by all messages.
 ///
 /// Messages sent to and received from Google Cloud services may be wrapped in
-/// [Any][crate::any::Any]. `Any` uses a `@type` field to encoding the type
+/// [Any][crate::any::Any]. `Any` uses a `@type` field to encode the type
 /// name and then validates extraction and insertion against this type.
 pub trait Message {
     /// The typename of this message.
     fn typename() -> &'static str;
 
+    /// Returns the serializer for this message type
     #[doc(hidden)]
-    /// Store the value into a JSON object.
-    fn to_map(&self) -> Result<Map, Error>
+    fn serializer() -> Box<dyn MessageSerializer<Self>>
     where
-        Self: serde::ser::Serialize + Sized,
+        Self: Sized + serde::ser::Serialize + serde::de::DeserializeOwned + 'static,
     {
-        to_json_object(self)
+        Box::new(DefaultSerializer::<Self>::new())
     }
+}
 
-    #[doc(hidden)]
-    /// Extract the value from a JSON object.
-    fn from_map(map: &Map) -> Result<Self, Error>
-    where
-        Self: serde::de::DeserializeOwned,
-    {
+/// Internal API for message serialization.
+/// This is not intended for direct use by consumers of this crate.
+#[doc(hidden)]
+pub trait MessageSerializer<T> {
+    fn to_map(&self, message: &T) -> Result<Map, Error>;
+    fn from_map(&self, map: &Map) -> Result<T, Error>;
```

Comment:
`clippy` does not like `from_*` when something accepts a `&self`. That is valid, but we are using `&self` and concrete serializer objects to workaround the fact that we can't default types in a trait.

So maybe we change these function names to `serialize(...)`, `deserialize(...)` ?

Reply:
> So maybe we change these function names to `serialize(...)`, `deserialize(...)` ?

Okay. I think that makes sense.

Reply:
I have gone with `serialize_to_map` and `deserialize_from_map` respectively.